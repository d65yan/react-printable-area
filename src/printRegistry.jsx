/**
 * @author Dayan Moreno
 * @fileOverview This component marks an area from which content can be printed
 * @module react-printable-area/PrintArea
 * @requires React
 *
 * @param {string} title - Page title to used on printed document
 * @param {string} printMode - [popup|iframe] if none is provided iframe will be used for all
 * browser but FireFox
 * @returns React Component
 *
 * @example
 * import PrintArea from 'react-printable-area/PrintArea';
 * import PrintSection from 'react-printable-area/PrintSection';
 * import PrintTrigger from 'react-printable-area/PrintTrigger';
 *
 * // in your render
 * <PrintArea>
 *   not printable content
 *   <PrintSection>
 *     printable content
 *   </PrintSection>
 *   <PrintTrigger><button>Print</button></PrintTrigger>
 * </PrintArea>
 *
 * @note
 * This component only controls printable content when PrintTrigger is used. If the browser's print
 * button is used all the print visible content of the page will be printed
 */

import { Component, createElement } from 'react';
import { string, shape, func } from 'prop-types';
import get from 'lodash/get';

export default wrapped => class extends Component {
  static displayName = `printRegistry(${get(wrapped, 'displayName', wrapped.name)})`;

  static wrappedComponent = wrapped;

  static propTypes = {
    className: string.isRequired,
  };

  sections = [];

  constructor(props) {
    super(props);
    this.state = {
      register: this.register.bind(this),
      deregister: this.deregister.bind(this),
      getHtml: this.getHtml.bind(this),
      setRef: this.setRef.bind(this),
    };
  }

  getSectionIndex = htmlGetter => this.sections.indexOf(
    this.sections.find(record => record.htmlGetter === htmlGetter)
  );

  getIsSectionRegistered = htmlGetter => this.getSectionIndex(htmlGetter) > -1;


  setRef = (printableRef) => { this.printableRef = printableRef; };

  getHtml() {
    const { className } = this.props;
    // eslint-disable-next-line unicorn/explicit-length-check
    if (!this.sections.length) {
      return (this.printableRef && this.printableRef.outerHTML) || '';
    }

    const htmlSections = this.sections.slice();
    htmlSections.sort((a, b) => {
      const nA = Number(a.priority || 0);
      const nB = Number(b.priority || 0);
      return nA === nB ? 0 : ((nB - nA) / Math.abs(nB - nA));
    });

    return `
      <div class="${className}">
        ${htmlSections.map(({ htmlGetter }) => htmlGetter()).join('\n\r')}
      </div>`;
  }

  deregister = (htmlGetter) => {
    const sectionIndex = this.getSectionIndex(htmlGetter);
    if (sectionIndex > -1) {
      this.sections.splice(sectionIndex, 1);
    }
  };

  register = (htmlGetter, priority) => (
    htmlGetter
    && !this.getIsSectionRegistered(htmlGetter)
    && this.sections.push({ htmlGetter, priority })
  );

  render() {
    return createElement(
      wrapped,
      {
        ...this.props,
        registry: this.state,
      }
    );
  }
};

export const registryShape = shape({
  register: func,
  deregister: func,
  getHtml: func,
  setRef: func,
});
