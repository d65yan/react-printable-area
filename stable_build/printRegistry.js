'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registryShape = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _propTypes = require('prop-types');

var _get = require('lodash/get');

var _get2 = _interopRequireDefault(_get);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author Dayan Moreno
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @fileOverview This component marks an area from which content can be printed
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @module react-printable-area/PrintArea
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @requires React
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @param {string} title - Page title to used on printed document
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @param {string} printMode - [popup|iframe] if none is provided iframe will be used for all
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * browser but FireFox
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @returns React Component
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @example
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * import PrintArea from 'react-printable-area/PrintArea';
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * import PrintSection from 'react-printable-area/PrintSection';
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * import PrintTrigger from 'react-printable-area/PrintTrigger';
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * // in your render
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * <PrintArea>
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *   not printable content
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *   <PrintSection>
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *     printable content
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *   </PrintSection>
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *   <PrintTrigger><button>Print</button></PrintTrigger>
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * </PrintArea>
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @note
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * This component only controls printable content when PrintTrigger is used. If the browser's print
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * button is used all the print visible content of the page will be printed
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

exports.default = function (wrapped) {
  var _class, _temp;

  return _temp = _class = function (_Component) {
    _inherits(_class, _Component);

    function _class(props) {
      _classCallCheck(this, _class);

      var _this = _possibleConstructorReturn(this, (_class.__proto__ || Object.getPrototypeOf(_class)).call(this, props));

      _this.sections = [];

      _this.getSectionIndex = function (htmlGetter) {
        return _this.sections.indexOf(_this.sections.find(function (record) {
          return record.htmlGetter === htmlGetter;
        }));
      };

      _this.getIsSectionRegistered = function (htmlGetter) {
        return _this.getSectionIndex(htmlGetter) > -1;
      };

      _this.setRef = function (printableRef) {
        _this.printableRef = printableRef;
      };

      _this.deregister = function (htmlGetter) {
        var sectionIndex = _this.getSectionIndex(htmlGetter);
        if (sectionIndex > -1) {
          _this.sections.splice(sectionIndex, 1);
        }
      };

      _this.register = function (htmlGetter, priority) {
        return htmlGetter && !_this.getIsSectionRegistered(htmlGetter) && _this.sections.push({ htmlGetter: htmlGetter, priority: priority });
      };

      _this.state = {
        register: _this.register.bind(_this),
        deregister: _this.deregister.bind(_this),
        getHtml: _this.getHtml.bind(_this),
        setRef: _this.setRef.bind(_this)
      };
      return _this;
    }

    _createClass(_class, [{
      key: 'getHtml',
      value: function getHtml() {
        var className = this.props.className;
        // eslint-disable-next-line unicorn/explicit-length-check

        if (!this.sections.length) {
          return this.printableRef && this.printableRef.outerHTML || '';
        }

        var htmlSections = this.sections.slice();
        htmlSections.sort(function (a, b) {
          var nA = Number(a.priority || 0);
          var nB = Number(b.priority || 0);
          return nA === nB ? 0 : (nB - nA) / Math.abs(nB - nA);
        });

        return '\n      <div class="' + className + '">\n        ' + htmlSections.map(function (_ref) {
          var htmlGetter = _ref.htmlGetter;
          return htmlGetter();
        }).join('\n\r') + '\n      </div>';
      }
    }, {
      key: 'render',
      value: function render() {
        return (0, _react.createElement)(wrapped, _extends({}, this.props, {
          registry: this.state
        }));
      }
    }]);

    return _class;
  }(_react.Component), _class.displayName = 'printRegistry(' + (0, _get2.default)(wrapped, 'displayName', wrapped.name) + ')', _class.wrappedComponent = wrapped, _class.propTypes = {
    className: _propTypes.string.isRequired
  }, _temp;
};

var registryShape = exports.registryShape = (0, _propTypes.shape)({
  register: _propTypes.func,
  deregister: _propTypes.func,
  getHtml: _propTypes.func,
  setRef: _propTypes.func
});