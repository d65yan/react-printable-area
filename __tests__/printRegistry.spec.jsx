import React from 'react';
import { shallow } from 'enzyme';
import printRegistry from '../src/printRegistry';

const Button = () => <button type="button" />;
const WrappedButton = printRegistry(Button);
const render = (props = {}) => shallow(<WrappedButton {...props} />);
describe('printRegistry', () => {
  it('is a function', () => {
    expect(printRegistry).toEqual(expect.any(Function));
  });

  it('Decorates the passed component', () => {
    expect(WrappedButton.wrappedComponent).toEqual(Button);
    expect(WrappedButton.displayName).toEqual('printRegistry(Button)');
  });

  it('passes the register props to the decorated component', () => {
    const wrappedButton = render();
    const button = wrappedButton.find(Button);
    expect(button.prop('registry')).toBe(wrappedButton.state());
    expect(wrappedButton.state()).toEqual({
      register: expect.any(Function),
      deregister: expect.any(Function),
      getHtml: expect.any(Function),
      setRef: expect.any(Function),
    });
  });

  it('passes the provided props to the decorated component', () => {
    const wrappedButton = render({ key1: 'value1' });
    const button = wrappedButton.find(Button);
    expect(button.prop('key1')).toEqual('value1');
  });

  describe('functionality', () => {
    it('register adds a section to the sections array if it does not exist yet', () => {
      const section1HtmlGetter = () => '1';
      const section2HtmlGetter = () => '2';

      const area = render();
      const instance = area.instance();
      instance.register(section1HtmlGetter, 5);
      instance.register(section2HtmlGetter, 1);
      instance.register(section1HtmlGetter, 0);

      expect(instance.sections).toHaveLength(2);
      expect(instance.sections).toEqual(
        [
          { htmlGetter: section1HtmlGetter, priority: 5 },
          { htmlGetter: section2HtmlGetter, priority: 1 },
        ]
      );
    });

    it('getSectionIndex returns the section index in the register array', () => {
      const section1HtmlGetter = () => '1';
      const section2HtmlGetter = () => '2';

      const area = render();
      const instance = area.instance();
      instance.register(section1HtmlGetter, 0);
      instance.register(section2HtmlGetter, 1);

      expect(instance.getSectionIndex(section2HtmlGetter)).toEqual(1);
    });

    it('getIsSectionRegistered returns true if the section has been registered', () => {
      const section1HtmlGetter = () => '1';
      const section2HtmlGetter = () => '2';

      const area = render();
      const instance = area.instance();
      instance.register(section1HtmlGetter);
      instance.register(section2HtmlGetter);

      expect(instance.getIsSectionRegistered(section2HtmlGetter)).toEqual(true);
    });

    it('deregister removes a section from the sections array if it does exist', () => {
      const section1HtmlGetter = () => '1';
      const section2HtmlGetter = () => '2';
      const section3HtmlGetter = () => '3';
      const section4HtmlGetter = () => '4';

      const area = render();
      const instance = area.instance();
      instance.register(section1HtmlGetter, 5);
      instance.register(section2HtmlGetter, 1);
      instance.register(section3HtmlGetter, 0);

      instance.deregister(section2HtmlGetter);
      expect(() => instance.deregister(section4HtmlGetter)).not.toThrow();
      expect(instance.sections).toHaveLength(2);
      expect(instance.sections).toEqual(
        [
          { htmlGetter: section1HtmlGetter, priority: 5 },
          { htmlGetter: section3HtmlGetter, priority: 0 },
        ]
      );
    });

    it('getRef sets the reference of the printable content', () => {
      const area = render();
      const instance = area.instance();
      instance.setRef('crazy ref');
      expect(instance.printableRef).toEqual('crazy ref');
    });

    describe('getHtml', () => {
      it('when no sections are registered and no ref returns empty string', () => {
        const registry = render();
        const registryProps = registry.find(Button).props().registry;
        expect(registryProps.getHtml()).toEqual('');
      });

      it('when no sections are registered returns outerHtml content of the ref', () => {
        const registry = render();
        const registryProps = registry.find(Button).props().registry;
        registryProps.setRef({ outerHTML: 'some content' });
        expect(registryProps.getHtml()).toEqual('some content');
      });

      it('when there are registered sections a combination of the getHtml will be returned', () => {
        const section1HtmlGetter = () => '<html section1>';
        const section2HtmlGetter = () => '<html section2>';
        const section3HtmlGetter = () => '<html section3>';

        const registry = render();
        const registryProps = registry.find(Button).props().registry;

        registryProps.register(section2HtmlGetter, 1);
        registryProps.register(section3HtmlGetter, 6);
        registryProps.register(section1HtmlGetter, 5);

        expect(registryProps.getHtml()).toMatchSnapshot();
      });

      it('when the registered sections have no priority they are presented as registered', () => {
        const section1HtmlGetter = () => '<html section1>';
        const section2HtmlGetter = () => '<html section2>';
        const section3HtmlGetter = () => '<html section3>';

        const registry = render();
        const registryProps = registry.find(Button).props().registry;

        registryProps.register(section2HtmlGetter);
        registryProps.register(section3HtmlGetter);
        registryProps.register(section1HtmlGetter);

        expect(registryProps.getHtml()).toMatchSnapshot();
      });
    });
  });
});
