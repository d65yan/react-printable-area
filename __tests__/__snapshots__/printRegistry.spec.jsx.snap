// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printRegistry functionality getHtml when the registered sections have no priority they are presented as registered 1`] = `
"
      <div class=\\"undefined\\">
        <html section2>

<html section3>

<html section1>
      </div>"
`;

exports[`printRegistry functionality getHtml when there are registered sections a combination of the getHtml will be returned 1`] = `
"
      <div class=\\"undefined\\">
        <html section3>

<html section1>

<html section2>
      </div>"
`;
